src/wallet/wallet.cpp

```diff
@@ -1868,8 +1868,18 @@ bool CWalletTx::IsTrusted() const
         if (parent == NULL)
             return false;
         const CTxOut& parentOut = parent->tx->vout[txin.prevout.n];
-        if (pwallet->IsMine(parentOut) != ISMINE_SPENDABLE)
-            return false;
+        const auto& isMine = pwallet->IsMine(parentOut);
+        if (isMine != ISMINE_SPENDABLE)
+        {
+            // If the wallet is hdwatchonly, check if it is a key we generated
+            if (!pwallet->IsHDWatchOnly() || isMine != ISMINE_WATCH_SOLVABLE)
+                return false;
+            const auto& meta = pwallet->mapKeyMetadata;
+            auto it = meta.find(CScriptID(parentOut.scriptPubKey));
+            if (it == meta.end() ||
+                it->second.hdKeypath.empty())
```

Comment:
I think we may need to generalize this check a bit more. If the script here is a multisig in which the watch-only wallet owns all the keys, this will return false since there is no hdKeypath.

Practically this means that spending 0-conf p2sh funds is a dicey experience. 

Perhaps something like `pwalletMain->IsHDWatchOnly(CScriptID& scriptID)` to make this check, which accounts for this situation?